### DISCUSSION TITLE
Godot Engine plugin and module related proposals

### DISCUSSION THREAD

**Discussion by Xrayez** [godot-extended-libraries/godot-ideas#46] [General]
This tracker aims to keep track of [Godot proposals](https://github.com/godotengine/godot-proposals/issues) which either: 

* can be implemented as a module/plugin;
* discussion needed before it can create demand by users (aka "nice-to-have" vs "need-to-have");
* use cases which are considered "too niche" to be implemented in core;
* [proposals targeted for core but which help to implement modules/plugins better](https://github.com/godotengine/godot-proposals/issues?q=is%3Aopen+is%3Aissue+label%3Atopic%3Aplugin).

Some items link to issues in this repository as well (to collect similar ideas under the same topic).

# Proposals

### Core
- [x] godotengine/godot-proposals#727 (see [Vector Resource](https://github.com/Xrayez/godot-vector-resource) plugin).
- [ ] godotengine/godot-proposals#2878

### Nodes
- [x] godotengine/godot-proposals#608
- [ ] godotengine/godot-proposals#725
- [ ] godotengine/godot#33610
- [ ] godotengine/godot-proposals#658
- [ ] godotengine/godot-proposals#960

### Input
- [ ] godotengine/godot-proposals#100.

### Gameplay/AI
- [x] godotengine/godot-proposals#281:
    - https://github.com/fire/godot-behavior-tree
- [ ] godotengine/godot-proposals#551.
- [ ] godotengine/godot#8847
- [ ] godotengine/godot#25021
- [ ] godotengine/godot-proposals#405

### Multimedia: video/audio/image
- [ ] godotengine/godot-proposals#33
- [ ] godotengine/godot-proposals#505
- [ ] godotengine/godot-proposals#543
- [ ] godotengine/godot-proposals#1037
- [x] godotengine/godot-proposals#2758:
    - https://github.com/2shady4u/godot-kra-psd-importer

### Debugging
- [x] godotengine/godot-proposals#115 ([added](godot-extended-libraries/godot-next#49)).
- [ ] godotengine/godot-proposals#112
- [ ] godotengine/godot-proposals#63

### Networking
- [ ] godotengine/godot-proposals#75:
    - https://github.com/SaracenOne/network_manager
    - https://github.com/Razzeeyy/godot-networked-instancing-example
    - https://github.com/Jummit/Godot-Replicator-Node-Plugin

### Physics
- [ ] godotengine/godot-proposals#116
- [ ] godotengine/godot-proposals#86
- [ ] godotengine/godot-proposals#756

### Editor/UI
- [x] godotengine/godot-proposals#69 (added to Godot 3.2).
- [ ] godotengine/godot-proposals#97
- [ ] godotengine/godot-proposals#13
- [ ] godotengine/godot-proposals#748
- [ ] godotengine/godot-proposals#2995

### Miscellaneous
- [x] godotengine/godot-proposals#965

---

See other [closed](https://github.com/godotengine/godot-proposals/issues?q=is%3Aissue+is%3Aclosed) or potentially [controversial](https://github.com/godotengine/godot-proposals/issues?q=is%3Aissue+sort%3Acomments-desc+is%3Aopen) proposals which may be worth to adding to this list.

**Comment by willnationsdev** [comment]
Proposal 65 could probably be omitted from the list imo. The Godot Editor itself is the feature that the OP there has proposed. They just didn't realize scene inheritance could be used that way. And there's no such thing as a generic solution for a game generator. You can make a template/framework for generating a particular type of game, but the open-ended, flexible version of what they want is what the Godot Editor is.

**Comment by Xrayez** [comment]
I think there could be multiple ways to do the same thing, especially for plugin candidates like this. I wanted to extract some potential idea I presented there. Created a section for this. 🙂

**Comment by golddotasksquestions** [comment]
The problem I have with this, is, that features asked for in those proposals are not substantially different from features that already are build into the engine, or those still getting added in future. 

Can someone explain to me how a the "Fake 3D effect for side scrolling platformers" is of any more benefit or value to all Godot users than a BehaviorTree class, DialogTree class, or InputBuffer class?

For people who _need_ those features you listed or have the opinion that they would make Godot a much more powerful tool out of the box, the decision to cast them aside as addon/plugin/module seems incredibly arbitrary if not to say like favoritism.

**Comment by Xrayez** [comment]
To be concrete and as an example: the reason I added godotengine/godot-proposals#281 is because there's already a WIP module being worked on: https://github.com/fire/godot-behavior-tree. It would be useful to at least encourage people to work on a *working* standalone implementation first before it can be considered for core, else it's just a waste of time to propose this stuff without any background. One could as well list all of the closed/controversial proposals but what would be the point of that?

There are certainly some proposals that don't belong to the core (like specific third-party integrations), maybe that's what this tracker should be about?

I don't think there's anything bad in listing proposals which *can* be implemented as a standalone solution, but it doesn't mean that it can't be added as part of Godot in the future: https://github.com/godotengine/godot/tree/master/modules. I guess the Bullet module is a good example for this.

The favoritism can be salvaged by people posting more links to this tracker, I'm the only one doing this at the moment. It's difficult to make objective decisions. 😛

What proposals do you think should be removed from the list, or added?

**Comment by willnationsdev** [comment]
@golddotasksquestions

I feel like the classes you've specified have a variety of ways in which they could be implemented and a variety of levels of depth and specialization that they could or could not have. Creating a single source of authority for them would therefore be more difficult to maintain as it would need to support many different avenues of work. The devs don't want to end up in a position where users are relying on addon stuff when there is already something for it in the engine. If that ever happens, it means they need to make changes to the features in the engine (to take over the advantages of that addon) or they need to strip the feature out of the engine entirely and let it be maintained in separate repos as an addon.

The best fix for this is to create a much healthier addon ecosystem that is easy to use and that's what I feel like Godot needs to focus on next. At least, that's where I want to concentrate my efforts, although most of it is out of my hands because core devs have ideas of how they'd like to implement it all themselves I bet. Just have to wait until they are free to work on it.

**Comment by DleanJeans** [comment]
Yes, we need an ecosystem of _official_ addons  so we can siphon our efforts into making quality addons instead of many creating their own versions with different levels of quality and inconsistent art style with the engine.

**Comment by Xrayez** [comment]
I've went through all current proposals now at GIP.

Around 60% are editor enhancement proposals, 25% are targeted for core stuff as it's not possible to implement them via script (both enhancements and new features), 5% are achieved as being invalid/not matching requirements (thanks Calinou for putting labels, it doesn't make much sense to track "invalid" proposals here just for the sake of it now).  We're are left with around 5-10% percent of proposals which are more or less valid candidates which can be implemented as a plugin.

That way, I don't see how the list can be arbitrary. I've add some more proposals to the list now grouped by categories, sorted alphabetically.

And yes, even if something can be implemented as a plugin, it doesn't mean that something is "put aside". Nobody is actually in charge here making decisions towards what can be implemented next, for me it's just a way to spend my time and tame my worried mind by organizing stuff from time to time. 🙂
